// unexpected-htmllike-raw-adapter/lib/RawAdapter.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var TYPEOF_IDENTIFIER = { $$typeof: 'raw json adapter' };

var baseObject = Object.create(Object, {
    $$typeof: {
        enumerable: false,
        writable: false,
        value: TYPEOF_IDENTIFIER
    }
});

function isRawType(value) {
    var type = typeof value;
    return type === 'string' || type === 'number' || type === 'boolean' || type === 'undefined' || value === null;
}

var DefaultOptions = {
    concatTextContent: false
};

function convertValueTypeToString(value) {

    if (typeof value === 'string') {
        // Common case can be fasttracked
        return value;
    }

    if (value === null || value === undefined) {
        return '';
    }

    return '' + value;
}

function concatenateStringChildren(accum, value) {
    if (isRawType(value) && accum.length && isRawType(accum[accum.length - 1])) {
        accum[accum.length - 1] = convertValueTypeToString(accum[accum.length - 1]) + convertValueTypeToString(value);
        return accum;
    }
    accum.push(value);
    return accum;
}

function flatten(value) {
    if (!Array.isArray(value)) {
        return [value];
    }

    return value.reduce(function (result, item) {
        return result.concat(flatten(item));
    }, []);
}

var RawAdapter = (function () {
    function RawAdapter(options) {
        _classCallCheck(this, RawAdapter);

        this._options = (0, _objectAssign2['default'])({}, DefaultOptions, options);
    }

    _createClass(RawAdapter, [{
        key: 'getName',
        value: function getName(element) {
            if (typeof element.type === 'string') {
                return element.type;
            }

            return 'no-display-name';
        }
    }, {
        key: 'getAttributes',
        value: function getAttributes(element) {

            return element.props;
        }
    }, {
        key: 'getChildren',
        value: function getChildren(element) {

            var children = element.children || [];
            children = children.filter(function (child) {
                return child !== null && child !== undefined;
            });
            if (this._options.convertToString || this._options.convertMultipleRawToStrings && children.length > 1) {
                children = children.map(function (child) {
                    return isRawType(child) ? convertValueTypeToString(child) : child;
                });
            }

            if (this._options.concatTextContent) {
                return children.reduce(concatenateStringChildren, []);
            }
            return children;
        }
    }, {
        key: 'setOptions',
        value: function setOptions(newOptions) {

            this._options = (0, _objectAssign2['default'])({}, this._options, newOptions);
        }
    }, {
        key: 'getOptions',
        value: function getOptions() {
            return this._options;
        }
    }, {
        key: 'serialize',
        value: function serialize(adapter, element) {
            var _this = this;

            if (isRawType(element)) {
                return element;
            }

            var serialized = Object.create(baseObject);
            serialized.type = adapter.getName(element);
            serialized.props = adapter.getAttributes(element);
            for (var prop in serialized.props) {
                if (typeof serialized.props[prop] === 'function') {
                    serialized.props[prop]._isRawDeserializedFunction = true;
                }
            }
            serialized.children = (adapter.getChildren(element) || []).map(function (child) {
                return _this.serialize(adapter, child);
            });
            return serialized;
        }
    }, {
        key: 'isRawElement',
        value: function isRawElement(value) {
            return value && typeof value === 'object' && value.$$typeof === TYPEOF_IDENTIFIER;
        }
    }, {
        key: 'deserialize',
        value: function deserialize(value) {
            // Actually, this method just converts a raw JSON-ish value to objects with the correct $$typeof property,
            // So it's exactly the same as just serializing using this adapter
            return this.serialize(this, value);
        }
    }]);

    return RawAdapter;
})();

RawAdapter.prototype.classAttributeName = 'className';

exports['default'] = RawAdapter;
module.exports = exports['default'];
